Implementation of TensorPad pass to align shared meme transpose operands

KoolJBlack opened the pull request:
`TensorPad` applies the `TransposePadOpPattern`. This pattern finds transpose linalg operations and pads their input/output values to the shared mem transpose tile size of 32x32. Right now this pass is only used in a test. Will be developed further in followup PRs. 

Limitations:
* This pads all operands (input/output) even if not necessary
* Only supports 2D transposes and assumes tile size of 32x32 (will be addressed in followup)

Part of  https://github.com/iree-org/iree/issues/10005

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/test/tensor_pad.mlir:
```diff
@@ -0,0 +1,66 @@
+//       CHECK:      ^bb0(%[[ARG2:.*]]: index, %[[ARG3:.*]]: index):
+//       CHECK:        tensor.yield %[[CST]] : f32
+//       CHECK:      } : tensor<32x?xf32> to tensor<32x32xf32>
+//       CHECK:      %[[D11:.*]] = tensor.pad %[[D7]] low{{\[}}%[[C0]], %[[C0]]] high{{\[}}%[[D9]], %[[C0]]] {
```
I thought we could avoid padding the output?

KoolJBlack replied:
I'm saving that for a followup. 

I'm also worried that always defaulting the output to an init tensor may be incorrect in certain situations, for instance if this linalg is fused with a producer/consumer that actually makes use of the accumulator?

A better solution might be to extend the pattern such that it looks up the source op to ensure it was created by an `init_tensor` before not using pad?

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Utils/MarkerUtils.h:
```diff
@@ -37,6 +37,9 @@ StringRef getWorkgroupL1TileMarker();
 StringRef getCopyToWorkgroupMemoryMarker();
 
+/// Marker for transpose padding operations.
+StringRef getTransposePadMarker();
```
if it is only used internally to the pass you don't need to add the marker here, it is better to keep it as a static in the pass

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,174 @@
+  return linalgOp.getNumParallelLoops() == 2;
+}
+
+struct TransposePadOpPattern : public OpRewritePattern<linalg::GenericOp> {
```
You don't really need a pattern here, I know we have been heavily relying on rewrite patterns combined with attributes but we are moving away from them and moving to applying direct transformation using APIs as you have above `FailureOr<SmallVector<Value>> rewriteAsPaddedOp`. This will ensure easier transition to transform dialect.
You should be able to simplify this code and remove the need for an attribute this way. You can just walk the function and apply the `rewriteAsPaddedOp` based on the precondition.

ThomasRaoux replied:
Using Rewriter is fine, and it can be done without using `OpRewritePattern`. The pattern rewrite will run to fix point which is really not what we want in this case so not using it should make a lot of things simpler.



> @rsuderman also advised against directly manipulating IR with complicated changes as unintended side effects persist if the pass fails/bails out early. I do prefer the error handling of the rewriter (its a nice bonus).

The bailing out of the rewrite pattern really doesn't buy you much as it doesn't clean up anything. You still need to first do all the precondition checks then do the transformation.

> I can't speak to the Transform Dialect ( I don't have any details on it). Can I get more context here? (we can discuss in person).

Yes that probably requires a long answer.

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,174 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/LinalgOpInfo.h"
+#include "iree/compiler/Codegen/LLVMGPU/TransposeUtils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Utils/MarkerUtils.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowTypes.h"
+#include "llvm/ADT/MapVector.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/SideEffectUtils.h"
```
nit: could you remove unneeded headers

MaheshRavishankar reviewed (dismissed):
Broad comment here. This is doing a lot of the same things that had already been tried upstream with respect to padding. If this is all happening at the tensor level this seems to be re-inveting a lot of things that have already been looked at.
@hanhanW who looked at it in quite a lot of detail.

ThomasRaoux commented:
> Broad comment here. This is doing a lot of the same things that had already been tried upstream with respect to padding. If this is all happening at the tensor level this seems to be re-inveting a lot of things that have already been looked at. @hanhanW who looked at it in quite a lot of detail.

What part in particular? I assume you are talking about `rewriteAsPaddedOp` as the rest is really just boilerplates. As far as I understand the upstream logic doesn't work out of the box which is why some logic is duplicated there. It would be good to unify that upstream but I'm not sure we have enough knowledge to do it at this point. 
What would you suggest?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    Value paddingValue = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(
+                 tensorLoad.getType().cast<ShapedType>().getElementType()));
```
Can also use `getElementTypeOrSelf(tensorLoad)`?

hanhanW commented:
> What part in particular? I assume you are talking about `rewriteAsPaddedOp` as the rest is really just boilerplates. As far as I understand the upstream logic doesn't work out of the box which is why some logic is duplicated there. It would be good to unify that upstream but I'm not sure we have enough knowledge to do it at this point. What would you suggest?

I think we should use the upstream method -- `rewriteAsPaddedOp`. I skim through two methods and found that the main difference is finding the bounding box. We can maybe passing a control function to the method.

Before doing that, I have a question. The upstream one looks into the producer chain until getting a slice op. It then extract the value if possible. I looked into the test IR, and would like to know if it's possible to generate similar IR that fits the upstream need.

```mlir
    %4 = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0)
    %5 = affine.apply affine_map<()[s0] -> (s0 * 32)>()[%workgroup_id_x]
    %6 = affine.apply affine_map<()[s0] -> (s0 * 32)>()[%workgroup_count_x]
    scf.for %arg1 = %5 to %c32 step %6 {
      %7 = flow.dispatch.tensor.load %1, offsets = [%arg0, %arg1], sizes = [%4, 32], strides = [1, 1] : !flow.dispatch.tensor<writeonly:48x32xf32> -> tensor<?x32xf32>
      %8 = flow.dispatch.tensor.load %0, offsets = [%arg1, %arg0], sizes = [32, %4], strides = [1, 1] : !flow.dispatch.tensor<readonly:32x48xf32> -> tensor<32x?xf32>
      %9 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d1, d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%8 : tensor<32x?xf32>) outs(%7 : tensor<?x32xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[32, 32]]>} {
      ^bb0(%arg2: f32, %arg3: f32):
        linalg.yield %arg2 : f32
      } -> tensor<?x32xf32>
      flow.dispatch.tensor.store %9, %1, offsets = [%arg0, %arg1], sizes = [%4, 32], strides = [1, 1] : tensor<?x32xf32> -> !flow.dispatch.tensor<writeonly:48x32xf32>
    }
  }
```

The operands of generic op are `%7` and `%8`. The unknown sizes are all `%4`. Instead of hard-coded `paddedShape = {32, 32};`, we should extract the value from `%4` right? It would be easier if we make the input IR like below. In this form, we'll be able to use upstream method.

```
%x = flow.dispatch.tensor.load %0 = [%arg0, %arg1] ... // full tensor
%y = tensor.extract_slice %x[0, 0][32, %4] ....
%z = linalg.generic .. ins(%y ..)
```

KoolJBlack commented:
> 

@hanhanW and I did discuss the upstream patterns. I tried using them directly as intended but they did not work out of the box. I scrapped the IR examples once I moved on to my own implementation so going from memory:

* upstream patterns don't accept our `IREE::Flow::DispatchTensorLoadOp`. instead they depend on extract slice for the input. 
* as HanHan mentioned, you can't pass padding shapes directly, only specify the dimensions to pad and the values to pad with. In the case of transpose, I wanted to pad directly to the tile size, currently hard coded to 32x32 for this simple example. But this will change in followup to support more dimensions for padding. 

Padding transpose is a bit different in the sense that the padding applies to specific dimensions of the transpose for each operand. I didn't see a way to make this distinction with the upstream pass, though its likely part of its bounding box calculation. The padded tile size must match the transpose tile size determined by the transpose order in https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp/#L535

Re @hanhanW question: the `%4` is the dimension of the tile that is incomplete due to the input not aligning exactly with the tile size. Normally we set the tile size to 1 for unaligned shapes, but I'm forcing it to be a fixed value and then using this code to pad up to it. You can see that its determined by ` %4 = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0)` which comes from the workgroup index and tile size. Not sure what you mean by:

> we should extract the value from %4 right?

Not sure about your second question either. The `flow.dispatch.tensor.load` are introduced by tiling to workgroup and they stick around until bufferization. If I replaced them with an extract slice just to run the upstream pattern, I suspect I'd need to do a similar transformation to get them back for subsequent passes?

I spent quite a bit of time digging through https://github.com/iree-org/iree-llvm-fork/blob/main/mlir/lib/Dialect/Linalg/Transforms/Transforms.cppto understand if this could be used, but there is prolly something I missed. The similarities born from some of the useful upstream functions that I am reusing, like `makeComposedPadHighOp`.  But ultimately the pattern seems to specific to a different focus for padding and not modular enough to reuse bits without extra transformations applies before and after it? I'd question the value of using them with the added code/complexity.

hanhanW commented:
> currently hard coded to 32x32 for this simple example. But this will change in followup to support more dimensions for padding.
> You can see that its determined by %4 = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0) which comes from the workgroup index and tile size.

We should not hardcoded the values.. The tiling size is 32. That's why we're seeing `%4 = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0)`. Basically, we should look into the affine.min op and extract the tiling value (i.e., 32 in this case) out. We should not just hard-code 32. There is a method for doing this -- `getConstantUpperBoundForIndex`.

To achieve what I said, we can look into the actual examples. I'm wondering if we can make the input IR

```mlir
%4 = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0)
%7 = flow.dispatch.tensor.load %1, offsets = [%arg0, %arg1], sizes = [%4, 32], strides = [1, 1] : !flow.dispatch.tensor<writeonly:48x32xf32> -> tensor<?x32xf32>
%8 = flow.dispatch.tensor.load %0, offsets = [%arg1, %arg0], sizes = [32, %4], strides = [1, 1] : !flow.dispatch.tensor<readonly:32x48xf32> -> tensor<32x?xf32>
%9 = linalg.generic {...} ins(%8 : tensor<32x?xf32>) outs(%7 : tensor<?x32xf32>)
```

to be

```mlir
%x = flow.dispatch.tensor.load %0 = [%arg0, %arg1] ... // full tensor
%y = flow.dispatch.tensor.load %0 = [%arg1, %arg0] ... // full tensor
%sz = affine.min affine_map<(d0) -> (-d0 + 48, 32)>(%arg0)
%sub_x = tensor.extract_slice %sub_x[0, 0][%4, 32] ....
%sub_y = tensor.extract_slice %sub_y[0, 0][32, %4] ....
%9 = linalg.generic {...} ins(%sub_y : tensor<32x?xf32>) outs(%sub_x : tensor<?x32xf32>)
```

If we can do it, the upstream version should work. The upstream version would get the slice, and call `getConstantUpperBoundForIndex` to get `32`.

I suspect we can move ConvertToDestinationPassingStylePass to an earlier stage, then we can get the IR I mentioned.

MaheshRavishankar commented:
I am not following all the details here, but if the issue is that it is looking for a `tensor.extract_slice` instead of a `flow.dispatch.tensor.load`, then we've actually hit this a few times now... One way forward would be to prioritize adding the `SubsetExtractOpInterface` that Nicolas has mentioned a few times needed in the fullness of time. We could help with creating this interface, with the `tensor.extract_slice` and `tensor.insert_slice` implementing this interface. We move the implementations upstream that are looking for `tensor.extract_slice` and `tensor.insert_slice` to use this interface instead.
We've hit this a couple of times. Instead of duplicating the upstream implementation to get around this, thats a better convergence point... I dont think we want to keep ownership of all of this code in IREE... we want to rely/build all such things upstream.
Also, eventually we will probably stop using `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` this way... so anything built on that is going to be an issue....

KoolJBlack commented:
Hanhan and Thomas and I just discussed this a bit. Several things going on here so I'll try to summarize and and answer your question: 

First off, I'll remove the hardcoded tile size and implement Hanhan's suggestion in https://github.com/iree-org/iree/pull/10655#issuecomment-1269026636. Based on the affine maps providing the sizes, this should extract the tile sizes already set distribute to workgroups. 

Regarding flow tensor load. I have a [gist of the IR](https://gist.github.com/KoolJBlack/bd7d52670887c16458ba328a13ba5096) from this work in-progress pipeline using the `TensorPadPass` after tileAndDistributeToWorkgroup: [addGPUTransposePassPipeline](https://github.com/KoolJBlack/iree/blob/73a57b423aba8f4046a5f0a7b92d73a75955c888/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp#L259)

From what I'm understand, with the latest changes to move Vectorization before Bufferization, we're left with `flow.dispatch.tensor.load/flow.dispatch.tensor.store` after distribute to workgroup. The extract slice ops aren't introduced until tiling on the GPU. The `TensorPadPass` is going in place of the existing `TesnorAllocPass` to handle the allocation and copying of data into shared memory, but with padding this time. According to Hanhan, upstream expects pure tiling to be applied before padding, but in we're doing it the other way around here. 

Even if this were addressed, I still have the separate issue noted above about the inflexibility to apply padding to select operands. In this simple example I have, the input IR is actually a 'mhlo.transpose" that takes in one operand:

```mlir
 %0 = "mhlo.transpose"(%ins) {permutation = dense<[1,0]> : tensor<2xi64>} : (tensor<32x48xf32>) -> tensor<48x32xf32>
```

The output shape persists as an `linalg.init_tensor`  in the IR past distribution to workgroups  until `ConvertToDestinationPassingStyle` which replaces it with a `flow.dispatch.tensor.load` from a hal subspan, hence the two loads when there is only one input shape.  In practice, this is loading form an empty tensor, padding is unnecessary: introduces an allocation, fill, then copy when in reality, only the allocation is necessary. We don't have a concrete solution to this. Hanhan has suggested using the upsteam [populatePadOpVectorizationPatterns] (https://github.com/llvm/llvm-project/blob/074fcec1eabfc992c46c95df215b1caf5cf58970/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1095) which will vectorize this and then hopefully optimize it out, but not sure if this can work for similar reasons I ran into with padding. 

Ideally I can get to a point where I can fully use the upstream pattern, though they handle way more than I need. At the moment, they're not applicable. For this change which is just a step towards handling non-aligned, I think it should be a concern of a separate PR down the line (and will definitely need more discussion on `SubsetExtractOpInterface`). I'm going to modify this pass further in following PRs. Hanhan is fine addressing the size calculation for now while we figure out other issues. 

@ThomasRaoux can chime in if I'm missing anything else?

MaheshRavishankar commented:
> Hanhan and Thomas and I just discussed this a bit. Several things going on here so I'll try to summarize and and answer your question:
> 
> First off, I'll remove the hardcoded tile size and implement Hanhan's suggestion in [#10655 (comment)](https://github.com/iree-org/iree/pull/10655#issuecomment-1269026636). Based on the affine maps providing the sizes, this should extract the tile sizes already set distribute to workgroups.
> 
> Regarding flow tensor load. I have a [gist of the IR](https://gist.github.com/KoolJBlack/bd7d52670887c16458ba328a13ba5096) from this work in-progress pipeline using the `TensorPadPass` after tileAndDistributeToWorkgroup: [addGPUTransposePassPipeline](https://github.com/KoolJBlack/iree/blob/73a57b423aba8f4046a5f0a7b92d73a75955c888/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp#L259)
> 
> From what I'm understand, with the latest changes to move Vectorization before Bufferization, we're left with `flow.dispatch.tensor.load/flow.dispatch.tensor.store` after distribute to workgroup. The extract slice ops aren't introduced until tiling on the GPU. The `TensorPadPass` is going in place of the existing `TesnorAllocPass` to handle the allocation and copying of data into shared memory, but with padding this time. According to Hanhan, upstream expects pure tiling to be applied before padding, but in we're doing it the other way around here.
> 
> Even if this were addressed, I still have the separate issue noted above about the inflexibility to apply padding to select operands. In this simple example I have, the input IR is actually a 'mhlo.transpose" that takes in one operand:
> 
> ```mlir
>  %0 = "mhlo.transpose"(%ins) {permutation = dense<[1,0]> : tensor<2xi64>} : (tensor<32x48xf32>) -> tensor<48x32xf32>
> ```
> 
> The output shape persists as an `linalg.init_tensor` in the IR past distribution to workgroups until `ConvertToDestinationPassingStyle` which replaces it with a `flow.dispatch.tensor.load` from a hal subspan, hence the two loads when there is only one input shape. In practice, this is loading form an empty tensor, padding is unnecessary: introduces an allocation, fill, then copy when in reality, only the allocation is necessary. We don't have a concrete solution to this.

The fact the bufferization does an allocation, fill and copy is an implementation detail.

 Hanhan has suggested using the upsteam [populatePadOpVectorizationPatterns] ([llvm/llvm-project@`074fcec`/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1095](https://github.com/llvm/llvm-project/blob/074fcec1eabfc992c46c95df215b1caf5cf58970/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1095)) which will vectorize this and then hopefully optimize it out, but not sure if this can work for similar reasons I ran into with padding.
> 

That would be ideal.... 

> Ideally I can get to a point where I can fully use the upstream pattern, though they handle way more than I need. At the moment, they're not applicable. For this change which is just a step towards handling non-aligned, I think it should be a concern of a separate PR down the line (and will definitely need more discussion on `SubsetExtractOpInterface`). I'm going to modify this pass further in following PRs. Hanhan is fine addressing the size calculation for now while we figure out other issues.
> 
> @ThomasRaoux can chime in if I'm missing anything else?

There is a lot going on here. My concern is that we are building a whole amount of tech that is in IREE. Instead of that we should be biasing on moving things so that we can use things from upstream directly. I see that having the `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` after tile + distribute is causing an issue. So we should prioritize getting to a stage away from that and re-prioritize work instead of replicating tech from upstream in IREE.

KoolJBlack commented:
Switched to `getConstantUpperBoundForIndex`.

ThomasRaoux commented:
> There is a lot going on here. My concern is that we are building a whole amount of tech that is in IREE. Instead of that we should be biasing on moving things so that we can use things from upstream directly. I see that having the `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` after tile + distribute is causing an issue. So we should prioritize getting to a stage away from that and re-prioritize work instead of replicating tech from upstream in IREE.

Do you mean that there is bunch of work being done by others around padding? This PR is duplicating a fairly small amount of logic. Since we are still in the exploring phase I would rather not upstream things just for this case. If you think there are other usages and supporting `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` is important for some other work it could make sense to tackle it now. If this is just related to this PR I don't think it is a good idea as I expect that there will be more back and forth coming up while we get padding working for the GPU.

MaheshRavishankar commented:
> > There is a lot going on here. My concern is that we are building a whole amount of tech that is in IREE. Instead of that we should be biasing on moving things so that we can use things from upstream directly. I see that having the `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` after tile + distribute is causing an issue. So we should prioritize getting to a stage away from that and re-prioritize work instead of replicating tech from upstream in IREE.
> 
> Do you mean that there is bunch of work being done by others around padding? This PR is duplicating a fairly small amount of logic. Since we are still in the exploring phase I would rather not upstream things just for this case. If you think there are other usages and supporting `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` is important for some other work it could make sense to tackle it now. If this is just related to this PR I don't think it is a good idea as I expect that there will be more back and forth coming up while we get padding working for the GPU.

I mean that there is a lot of tech being built in IREE with no clear plan of upstreaming it. The divergence point seems to be the difference between IREE using `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` instead of `tensor.extract_slice` and `tensor.insert_slice`. (I think this has come up before). If this is a repeating occurence, I would rather prioritize moving to using `scf.foreach_thread` (and by proxy use `tensor.extract_slice`/`tensor.insert_slice`, etc.) and build/iterate over these things upstream. Duplicating the logic/independently evolving such things invariable make these IREE specific and harder to generalize. Is this really a blocker in terms of performance to achieve on models and driven by that?

MaheshRavishankar commented:
> > I mean that there is a lot of tech being built in IREE with no clear plan of upstreaming it. The divergence point seems to be the difference between IREE using `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` instead of `tensor.extract_slice` and `tensor.insert_slice`. (I think this has come up before). If this is a repeating occurence, I would rather prioritize moving to using `scf.foreach_thread` (and by proxy use `tensor.extract_slice`/`tensor.insert_slice`, etc.) and build/iterate over these things upstream.
> 
> Here `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` is one difference but for experimentation it is convenient to not be bound by the opinion of the upstream pattern. Regarding `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store`, scf.foreach_thread would solve it - as you mentioned - which is why I'm not a big fan about adding special support upstream to handle.
> 
> > Duplicating the logic/independently evolving such things invariable make these IREE specific and harder to generalize. Is this really a blocker in terms of performance to achieve on models and driven by that?
> 
> Yes, this is also needed sometime to be able to experiment. It's not a blocker at the time but supporting padding for the GPU is part of the roadmap and is an important piece of generalizing the performance of the compiler.
> 
> To summarize based on the discussion I see different solutions:
> 
> 1. Not land this PR and have Kojo continue the work on his own branch until everything is put together and we have more confidence on how padding should be handled
> 2. Re-evaluate the roadmap and decide that we need `scf.foreach_thread` transition before doing the padding work. Then we need to make sure we have a timeline for `scf.foreach_thread` and re-evaluate what we are working on at the time on the GPU side.

This should be part of the roadmap while planning GPU work as well. Outside of the GPU related work, this transition is not a blocker. If there was work like padding, (specialization from earlier, and there was one more) that all would benefit from moving to `scf.foreach_thread`, then that should have been part of the roadmap to start with? It isnt on my radar cause it is not a blocker for things that I am looking at right now, so I dont have this prioritized, planned etc. I dont know the details of padding implementation upstream to make the link that this is needed as a precursor.

> 3. Adding more control upstream to handle `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` but it feels like this is a side-effect of out current implementation.

I am not sure I follow what more control upstream implies. Broadly `tensor.extract_slice`/`tensor.insert_slice` are very similar in function as `flow.dispatch.tensor.load`/flow.dispatch.tensor.store` (at least AFAICS for the analysis here, but there are also some very real differences w.r.t side-effects). Upstream implementations are very tied to `tensor.extract_slice`/`tensor.insert_slice`. I think that would benefit with moving to using `SubsetExtractOpInterface` (which btw, doesnt exist, so if this was needed should have been prioritized appropriately based on discussion with Nicolas). Then the analysis/transformation would work without moving to `scf.foreach_thread`. So between (2) and (3) we could have picked one way forward.

> 4. Any other changes that would make this PR acceptable?
> 
> Did I miss any other solution? What do you think is the path forward?

We seem to have painted ourselves into a corner. I am very concerned that we will end up building tech islands that we wont be able to get out of. Without building things upstream in "a clean environment" to start with, it is inevitable that we will have some shortcuts, assumptions baked in. Given Kojo has already spent time working on this, of course I don't want it to go to waste, but seeing a lot of things being built "for CUDA" based on the current state of the common parts will just result in divergence from other backends that we might not be able to reconcile. We might/will end up in a situation where all backends end up in a state where progress on one is hampered due to issues with other backends (explicitly a non-goal for IREE). In all these cases the preferred way forward is
1) Adapt/generalize upstream implementation, and/or
2) Adapt IREE to allow using upstream implementation.
Least desirable is
- Copy/mimic implementation from upstream to get around representational issue due to expediency (we are explicitly telling other folks to not do this, so us doing it is a bit hypocritical).

Having raised the issue, I am not going to delve into implementation aspects here. I really hope we can avoid such situations with better ordering of tasks.

ThomasRaoux commented:
> We seem to have painted ourselves into a corner. I am very concerned that we will end up building tech islands that we wont be able to get out of. Without building things upstream in "a clean environment" to start with, it is inevitable that we will have some shortcuts, assumptions baked in. Given Kojo has already spent time working on this, of course I don't want it to go to waste, but seeing a lot of things being built "for CUDA" based on the current state of the common parts will just result in divergence from other backends that we might not be able to reconcile. We might/will end up in a situation where all backends end up in a state where progress on one is hampered due to issues with other backends (explicitly a non-goal for IREE). In all these cases the preferred way forward is
> 
> 1. Adapt/generalize upstream implementation, and/or
> 2. Adapt IREE to allow using upstream implementation.
> 
> Least desirable is
> 
> * Copy/mimic implementation from upstream to get around representational issue due to expediency (we are explicitly telling other folks to not do this, so us doing it is a bit hypocritical).
> 
> Having raised the issue, I am not going to delve into implementation aspects here. I really hope we can avoid such situations with better ordering of tasks.

The logic duplicated is ~50 lines of code and we have been doing this kind of work in IREE that we later generalize and upstream in the past. 
Therefore I assume your concern goes much beyond just the scope of this PR. I think we need to rework the expectations we have with regard to the work going in IREE. It sounds like you have serious concerns about the state of the backend. Let's solve that offline so that we don't bump into that at every PR.

MaheshRavishankar commented:
> > We seem to have painted ourselves into a corner. I am very concerned that we will end up building tech islands that we wont be able to get out of. Without building things upstream in "a clean environment" to start with, it is inevitable that we will have some shortcuts, assumptions baked in. Given Kojo has already spent time working on this, of course I don't want it to go to waste, but seeing a lot of things being built "for CUDA" based on the current state of the common parts will just result in divergence from other backends that we might not be able to reconcile. We might/will end up in a situation where all backends end up in a state where progress on one is hampered due to issues with other backends (explicitly a non-goal for IREE). In all these cases the preferred way forward is
> > 
> > 1. Adapt/generalize upstream implementation, and/or
> > 2. Adapt IREE to allow using upstream implementation.
> > 
> > Least desirable is
> > 
> > * Copy/mimic implementation from upstream to get around representational issue due to expediency (we are explicitly telling other folks to not do this, so us doing it is a bit hypocritical).
> > 
> > Having raised the issue, I am not going to delve into implementation aspects here. I really hope we can avoid such situations with better ordering of tasks.
> 
> The logic duplicated is ~50 lines of code and we have been doing this kind of work in IREE that we later generalize and upstream in the past. Therefore I assume your concern goes much beyond just the scope of this PR. I think we need to rework the expectations we have with regard to the work going in IREE. It sounds like you have serious concerns about the state of the backend. Let's solve that offline so that we don't bump into that at every PR.

Sounds good. Thats why I removed my blocker. I am not saying there is a dire need to fix here. Its more about a death by a thousand cuts (50 lines of code here, 100 lines of code there... these add up). Mostly looking to nip it in the bud. Happy to discuss offline.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,178 @@
+  // Set IP after op because we also take the dims of the original output.
+  rewriter.setInsertionPointAfter(linalgOp);
+
+  // Pad each input operand in shared memory out to 32,32
```
the comment is outdated

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,178 @@
+    ArrayRef<int64_t> shape = linalgOp.getShape(opOperand);
+    SmallVector<int64_t> paddedShape(shape.begin(), shape.end());
+    for (const auto &en : llvm::enumerate(tensorLoad.getMixedSizes())) {
+      if (en.value().is<Attribute>()) {
+        paddedShape[en.index()] =
+            en.value().get<Attribute>().dyn_cast<IntegerAttr>().getInt();
+      } else {
```
The check of dyn_cast<IntegerAttr> is missing, maybe use `getConstantIntValue` method, then we don't need to check ourself?

```mlir
Optional<int64_t> cst = getConstantIntValue(en.value());
...
```

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -47,8 +49,12 @@ static FailureOr<SmallVector<Value>> rewriteAsPaddedOp(
+              linalgOp, "No constant bounding box can be found for padding");
+        }
+        paddedShape[en.index()] = cst.getValue();
       } else {
```
We don't need the first check (i.e., is<Attribute>) if we're using getConstantIntValue method, right?

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp:
```diff
@@ -0,0 +1,154 @@
+      rewriter.replaceOp(linalgOp, *newResults);
+    });
+
+    // Remove all the markers at the end.
+    funcOp->walk([&](linalg::LinalgOp op) {
+      op->removeAttr(linalg::LinalgTransforms::kLinalgTransformMarker);
+    });
```
I don't think you need that anymore

KoolJBlack replied:
removed
